### PR TITLE
overall cleanup

### PR DESCRIPTION
**Pull request by Alireza9476**
### **New dependencies:**
Lombok
Apache Lang3

### **Refactoring:**
Deleted boilderplate code with the use of lombok.
Replaced imperative code by functional, imo looks better and cleaner

### **New:**
Created OwnerService (feel free to write fancy code inside)
Moved input validation to new Component
Added validation for new visit +  test cases.
Added seperate logik and message for date validation (empty and is in future)
Added some spaces between input items
Added new design for button hover (will revert, if it does not fit)
Wrote test methods for nearly every Entity

### CODE REVIEW

File: src/main/java/org/springframework/samples/petclinic/model/Person.java [hunk 1]
```diff
@@ -35,20 +44,24 @@ public class Person extends BaseEntity {
 	@NotBlank
 	private String lastName;
 
-	public String getFirstName() {
-		return this.firstName;
-	}
+	@Override
+	public boolean equals(Object object) {
```
Comment (Alireza9476) [inline src/main/java/org/springframework/samples/petclinic/model/Person.java:48]: Equals used in serialization test

File: src/main/java/org/springframework/samples/petclinic/owner/Owner.java [hunk 2]
```diff
@@ -59,43 +64,18 @@ public class Owner extends Person {
...
-	}
-
-	public List<Pet> getPets() {
-		return this.pets;
-	}
-
 	public void addPet(Pet pet) {
-		if (pet.isNew()) {
-			getPets().add(pet);
+		var petIsNewInList = Optional.of(this.pets).stream().anyMatch(value -> value.equals(pet));
```
Comment (Alireza9476) [inline src/main/java/org/springframework/samples/petclinic/owner/Owner.java:74]: A test case was messed up, and imo this new conditions makes sense besides the logic of isNewEntry

File: src/main/scss/petclinic.scss [hunk 3]
```diff
@@ -47,6 +47,14 @@ $pagination-active-bg:                    $spring-brown;
 $pagination-active-border:                $spring-green;
 $table-border-color:                      $spring-brown;
 
+$asterisk: "\f069";
+
+.asterisk{
```
Comment (Alireza9476) [inline src/main/scss/petclinic.scss:52]: ik there are existing asterisk classes but somehow didn't worked for me

File: src/test/java/org/springframework/samples/petclinic/owner/OwnerControllerTests.java [hunk 4]
```diff
@@ -57,47 +62,36 @@
...
 	@Autowired
 	private MockMvc mockMvc;
 
 	@MockBean
 	private OwnerRepository owners;
 
+	@MockBean
+	private OwnerService ownerService;
+
+	private record TestData(int ownerId, String ownerFirstname, String ownerLastname, String petName, PetType petType) {
```
Comment (Alireza9476) [inline src/test/java/org/springframework/samples/petclinic/owner/OwnerControllerTests.java:74]: Yeah I could create 5 private static variables but was too ugly

File: src/test/java/org/springframework/samples/petclinic/system/CrashControllerTests.java [hunk 5]
```diff
@@ -31,16 +32,12 @@
...
 	@Test
 	void testTriggerException() throws Exception {
-		RuntimeException thrown = assertThrows(RuntimeException.class, () -> {
-			testee.triggerException();
-		});
-
-		assertEquals("Expected: controller used to showcase what happens when an exception is thrown",
-				thrown.getMessage());
+		assertThatThrownBy(() -> crashController.triggerException()).isInstanceOf(RuntimeException.class)
+			.hasMessage("Expected: controller used to showcase what happens when an exception is thrown");
```
Comment (Alireza9476) [inline src/test/java/org/springframework/samples/petclinic/system/CrashControllerTests.java:40]: Way prettier

File: src/main/java/org/springframework/samples/petclinic/Pet/PetTypes.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+	CAT("cat", "pet.cat"), DOG("dog", "pet.dog"), LIZARD("lizard", "pet.lizard"), SNAKE("snake", "pet.snake"),
+	BIRD("bird", "pet.bird"), HAMSTER("hamster", "hamster.bird");
+
+	private final String value;
+
+	private final String frontendTranslationKey;
+
+	// TODO: Refactor so we will use this instead of taking all Types from db
+	public static List<String> getPetTypeNames() {
```
Comment (Alireza9476) [inline src/main/java/org/springframework/samples/petclinic/Pet/PetTypes.java:22]: I forgot to use the translation keys in frontend, but can be a TODO for another PR.

File: src/main/java/org/springframework/samples/petclinic/Pet/PetTypes.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+	CAT("cat", "pet.cat"), DOG("dog", "pet.dog"), LIZARD("lizard", "pet.lizard"), SNAKE("snake", "pet.snake"),
+	BIRD("bird", "pet.bird"), HAMSTER("hamster", "pet.hamster");
+
+	private final String value;
+
+	private final String frontendTranslationKey;
+
+	// TODO: Refactor so we will use this instead of taking all Types from db
+	public static List<String> getPetTypeNames() {
```
Comment (Alireza9476) [inline src/main/java/org/springframework/samples/petclinic/Pet/PetTypes.java:22]: Forgot to use the translation keys in frontend, can be a TODO from another PR

File: readme.md [hunk 8]
```diff
@@ -17,6 +17,9 @@ git clone https://github.com/spring-projects/spring-petclinic.git
 cd spring-petclinic
 ./mvnw package
 java -jar target/*.jar
+
+Extras:
+Also ,,enable annotation processing" in `Compiler > Annotation Processors` so that you can work with lombok properly
```
Comment (andrzejsydor) [inline readme.md:22]:
```suggestion
Also ,"enable annotation processing" in `Compiler > Annotation Processors` so that you can work with lombok properly 
```

**Comment by dsyer** [comment]
Thanks, but we don't want to use Lombok or Apache Lang.
